torch/autograd/functional.py

```diff
@@ -0,0 +1,712 @@
+        for j in range(out.nelement()):
+            # Do this instead of the grad_output trick in `gradcheck` to avoid version counter issues
+            vj = _autograd_grad(out.reshape(-1)[j], inputs, retain_graph=True, create_graph=create_graph)
+
+            for el_idx, (jac_i_el, vj_el, inp_el) in enumerate(zip(jac_i, vj, inputs)):
+                if vj_el is not None:
+                    if strict and create_graph and not vj_el.requires_grad:
+                        raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                           "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+                    jac_i_el.append(vj_el)
+                else:
+                    if strict:
+                        raise RuntimeError("Output {} of the user-provided function is independent of "
+                                           "input {}. This is not allowed in strict mode.".format(i, el_idx))
+                    jac_i_el.append(torch.zeros_like(inp_el))
+
+        jacobian += (tuple(torch.stack(jac_i_el, dim=0).view(out.size()
+                     + inputs[el_idx].size()) for (el_idx, jac_i_el) in enumerate(jac_i)), )
+
+    jacobian = _grad_postprocess(jacobian, create_graph)
+
+    return _tuple_postprocess(jacobian, (tuple_outputs, tuple_inputs))
+
+
+def hessian(func, inputs, create_graph=False, strict=False):
+    r"""Function that computes the Hessian of a given scalar function.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a Tensor with a single element.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        create_graph (bool, optional): If ``True``, the Hessian will be computed in
+            a differentiable manner. Note that when strict is ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
+            said inputs, which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        Hessian (Tensor or a tuple of tuple of Tensor) if there are a single input,
+            this will be a single Tensor containing the Hessian for the input.
+            If it is a tuple, then the Hessian will be a tuple of tuples where
+            ``Hessian[i][j]`` will contain the Hessian of the ``i``th input
+            and ``j``th input with size the sum of the size of the ``i``th input plus
+            the size of the ``j``th input.
+
+    Example::
+
+        >>> def pow_reducer(x):
+        ...   return x.pow(3).sum()
+        >>> inputs = torch.rand(2, 2)
+        >>> hessian(pow_reducer, inputs)
```

Comment:
It might be nicer to send in `inputs` as a tensor with one dimension so that the output comes out as a matrix.

Also, I've noticed that a lot of the examples use random inputs. Would it be better if they didn't use random inputs (and we instead did `inputs = torch.tensor([1., 2.])` for example) so that the user can "check" that the output matches their definition of hessian?

Reply:
After chatting with @gchanan I will postpone this to a later PR.